src/index.ts

```diff
@@ -236,82 +232,121 @@ async function catchupAndWatchChain(
     });
 
     chainLogger.info("catching up with blockchain events");
-    const catchupSentinel = new AsyncSentinel();
 
     const indexerLogger = chainLogger.child({ subsystem: "DataUpdater" });
 
     // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call
     const throttledLogProgress = throttle(
       5000,
       (currentBlock: number, lastBlock: number, pendingEventsCount: number) => {
+        const progressPercentage = ((currentBlock / lastBlock) * 100).toFixed(
+          1
+        );
+
         indexerLogger.info(
-          `pending events: ${pendingEventsCount} (indexing blocks ${currentBlock}-${lastBlock})`
+          `${currentBlock}/${lastBlock} indexed (${progressPercentage}%) (pending events: ${pendingEventsCount})`
         );
       }
     );
 
-    const indexer = await createIndexer(
-      rpcProvider,
-      storage,
-      async (indexer: Indexer<JsonStorage>, event: ChainsauceEvent) => {
-        try {
-          return await handleEvent(event, {
-            chainId: config.chain.id,
-            db: storage,
-            subscribe: (...args) => indexer.subscribe(...args),
-            ipfsGet: cachedIpfsGet,
-            priceProvider,
-            logger: indexerLogger,
-          });
-        } catch (err) {
-          indexerLogger.warn({
-            msg: "skipping event due to error while processing",
-            err,
-            event,
-          });
+    const eventHandlerContext: EventHandlerContext = {
+      chainId: config.chain.id,
+      db: storage,
+      ipfsGet: cachedIpfsGet,
+      priceProvider,
+      logger: indexerLogger,
+    };
+
+    // the chainsauce cache is used to cache events and contract reads
+    const chainsauceCache = config.cacheDir
+      ? createSqliteCache(path.join(config.cacheDir, "chainsauceCache.db"))
+      : null;
+
+    // the subscription store is used to keep track of indexed events for resuming
+    // indexing after a restart, it goes with the chain data
+    const subscriptionStore = createSqliteSubscriptionStore(
+      path.join(CHAIN_DIR_PATH, "subscriptions.db")
+    );
+
+    const indexer = createIndexer({
+      contracts: abis,
```

Comment:
If the `contracts` parameter here is about definitions rather than deployments, it might be worth to name that `abis` (or something similarly unambiguous), too?

Reply:
yeah you're right, naming it `abis` will make the separation clearer 👍

Reply:
in the end I kept the name `contracts` for this PR because it's not a simple rename here and needs more thought , because the name here affects other parts of the interface, I was thinking this could be an option but I am still not 100% sure:

```
import RoundImplementationV2  from "./abis";

const indexer = createIndexer({
    abis: {
       RoundImplementation
    },
   ...
});

indexer.subscribeToContract({
   abiName: "RoundImplementation",
   address: "0x...",
});

indexer.readContract({
    abiName: "RoundImplementation",
    address: "0x...",
    functionName: "roundMetaPtr"
});
```

so the way the interface is now the ABI is abstracted away after indexer creation and you only deal with "contracts":

1. you define your contracts (which are just the ABIs)
2. you subscribe to a deployed contract referencing the contract name
3. event handlers read from deployed contracts referencing the contract name

renaming contracts to ABIs:

1. you define your ABIs
2. you subscribe to a deployed contract referencing the ABI name
3. event handlers read from deployed contracts referencing the ABI name

I feel like calling them ABIs is the right name in the `createIndexer` call but down the line you're dealing with contracts and the ABI is an "implementation detail".

Reply:
resolving the conversation to merge, but open to revisit this later